Operational address translation and NAPOT PTEs

daniellustig opened the pull request:
Creating a local pull request for this branch so that people can add comments.

daniellustig committed:
More operational definition of address translation

This makes it easier to describe the behavior of NAPOT PTEs, and also
brings the text more in line with the operational Sail model.

daniellustig committed:
Update front matter

daniellustig committed:
Add support for NAPOT PTEs

gfavor commented on src/supervisor.tex:
```diff
@@ -1006,16 +1007,12 @@ \subsection{Supervisor Memory-Management Fence Instruction}
 \begin{commentary}
-The SFENCE.VMA is used to flush any local hardware caches related to
-address translation.  It is specified as a fence rather than a TLB
-flush to provide cleaner semantics with respect to which instructions
```
It seems like this sentence should be retained.  As non-normative text it explains some of the reason why this TLB maintenance operation is defined as a form of fence - which is useful for future readers of the spec to understand why this isn't defined as a "TLB invalidation".

But I now realize that you are trying to remove all mention of translation caching here - which is "hiding the ball" behind the more abstracted last sentence - leaving it as an exercise for the reader to realize that this instruction is about flushing TLB caches.  As a more operational style description this seems particularly undesirable.

daniellustig replied:
I removed the first sentence only because it seemed redundant with the sentence I added at line 1010.  I dropped the second sentence because it no longer seemed relevant after now defining SFENCE to actually flush the address-translation cache(s).  I don't really mind if we put these back, but the bit about "rather than a TLB flush" seems slightly awkward now given the new definition of SFENCE, so I'd still want to rewrite that at least.

aswaterman replied:
I could go either way on this point, but I certainly agree "the specified as a fence rather than a TLB flush" part doesn't belong in an operational description.

gfavor commented on src/supervisor.tex:
```diff
@@ -1042,22 +1039,21 @@ \subsection{Supervisor Memory-Management Fence Instruction}
+\item If {\em rs1}={\tt x0} and {\em rs2}={\tt x0}, the fence invalidates
+      all address-translation cache entries, for all address spaces.
+\item If {\em rs1}={\tt x0} and {\em rs2}$\neq${\tt x0}, the fence invalidates
+      all address-translation cache entries pointing to page table entries that
```
The terminology "cache entries **pointing** to page table entries" doesn't literally seem to be correct.  I understand the intent, but TLB entries generally don't "point to" PTEs.  Most commonly they, if anything, point to the final address translation result (although I wouldn't use the word "point" in this case either).

Plus a page walk cache entry only caches the first part of an overall tablework translation (for the sake of short-cutting and hence accelerating tablewalks), so such entries often times won't in any way point to a leaf PTE (they point to the next level page table to be walked).  This is part of why the old text made reference to "any level of the page tables".

You ultimately need text that covers both classic TLB entries as well as page walk (aka partial translation) caches.  Maybe say "invalidates all address-translation cache entries for non-global mappings (see Section ...) and that match the address space ..."

daniellustig replied:
I agree about not using the word "pointing", but I'll quibble with the claim we need to cover TLBs specifically.  The model is defined only in terms of this address-translation cache concept, which abstractly represents both TLBs and page walk caches.

In any case, something as simple as replacing "pointing to" with "derived from" should do the trick.  I don't know that we need to add even more language than that.

gfavor commented on src/supervisor.tex:
```diff
@@ -1042,22 +1039,21 @@ \subsection{Supervisor Memory-Management Fence Instruction}
+      are non-global and that match the address space identified by integer
+      register {\em rs2}.
+\item If {\em rs1}$\neq${\tt x0} and {\em rs2}={\tt x0}, the fence invalidates
+      all address-translation cache entries pointing to leaf page table entries
```
Similar adjustment to the terminology is desired here as well.  Maybe say "... cache entries that cache any part of the tablewalk necessary to translate the virtual address in ...."  Or say "... cache entries that cache any part of the overall translation of the virtual address in ...."  The key point is to cover partial as well as full translation caches.

daniellustig replied:
Resolving this comment in favor of keeping the discussion in one single comment above.  I'll make sure this whole block of text is updated in the end.

gfavor commented on src/supervisor.tex:
```diff
@@ -1042,22 +1039,21 @@ \subsection{Supervisor Memory-Management Fence Instruction}
-      space identified by integer register {\em rs2}.
-      Accesses to global mappings are not ordered.
+      invalidates all address-translation cache entries pointing to non-global
+      leaf page table entries corresponding to the virtual address in {\em
```
Similar adjustments (as above) to the terminology are desired here as well.

daniellustig replied:
Ditto prior comment

gfavor commented on src/supervisor.tex:
```diff
@@ -1447,10 +1448,18 @@ \subsection{Virtual Address Translation Process}
-  $a+va.vpn[i]\times \textrm{PTESIZE}$. (For Sv32, PTESIZE=4.)
-  If accessing $pte$ violates a PMA or PMP check, raise an
-  access-fault exception corresponding to the original access type.
+\item Let $b=a+va.vpn[i]\times \textrm{PTESIZE}$.  Either:
```
"Then either", at least to me , sounds better than "Either".  (Or alternatively remove the first word "Let" in each of the two items, and change "take" to "takes".  But the first suggestion is preferable.)

gfavor commented on src/supervisor.tex:
```diff
@@ -1447,10 +1448,18 @@ \subsection{Virtual Address Translation Process}
+      matches the address space identified by ${\tt satp}.ASID$, or $pte.g=1$.
+    \item Let $pte$ be the value of the PTE at address $b$.  If accessing $pte$
+      violates a PMA or PMP check, raise an access-fault exception
+      corresponding to the original access type.  Otherwise, create an
```
Note that "creating a cache entry" is not a requirement here; just the likely thing to happen

daniellustig replied:
From the point of view of the model, I don't think it makes a difference.  The translation algorithm permits an address-translation cache entry to be ignored.  So unless I'm missing something, specifying that the algorithm always adds the entry just eliminates one small and unnecessary degree of freedom.

gfavor commented on src/supervisor.tex:
```diff
@@ -1447,10 +1448,18 @@ \subsection{Virtual Address Translation Process}
+      address space identified in ${\tt satp}.ASID$.
+  \end{itemize}
+  (For Sv32, PTESIZE=4.)
 
```
It might be good here to have a non-normative comment clueing people into what is hidden behind the ball (of the first item), i.e. that a cache entry may map a larger region of memory - because of a NAPOT PTE - than would conventionally be implied by walking down to a 4K page table entry.

daniellustig replied:
We can add some quick commentary, but I'll point out that this part isn't just about NAPOT PTEs.  Any previously reachable PTE can be cached if the hart hasn't issued an SFENCE.

gfavor commented on src/supervisor.tex:
```diff
@@ -1625,8 +1634,21 @@ \subsection{Addressing and Memory Protection}
+Bits 62--60 indicate cacheability attributes associated with the virtual
+address in question, as described in Section~\ref{virtualattributes}.
+
+Bit 59 indicates the use of naturally aligned power-of-two (NAPOT)
```
Maybe swap bit 62-60 and bit 59.  This allows more convenient room for the attribute field to potentially grow in the future.

gfavor commented on src/supervisor.tex:
```diff
@@ -1650,6 +1672,108 @@ \subsection{Addressing and Memory Protection}
+\label{napotpages}
+
+Bit 59 indicates that a PTE represents a translation that is part of a range of
+contiguous virtual-to-physical translations with the same permissions at some
```
And same G, U, and V bits (as well as RWX bits)!  Essentially PTE bits 5-0.

daniellustig replied:
Right; will fix.

gfavor commented on src/supervisor.tex:
```diff
@@ -1650,6 +1672,108 @@ \subsection{Addressing and Memory Protection}
+\caption{NAPOT Contiguous Translation Encodings}
+\label{ptenapot}
+\end{table*}
+
```
If NAPOT support is implemented, must all NAPOT region sizes be implemented?  Or, if not, then what is the behavior for unsupported region sizes, e.g. a mapping size smaller than the specified NAPOT size may be cached.

More generally, in non-normative text, it may be worth noting that NAPOT encodings are hints to enable caching larger-than-4KiB mapping sizes, but any size less than or equal to the NAPOT region size may be cached.

daniellustig replied:
To me, the first question is still something we have to resolve as a group, so I didn't write much about it yet.

I'll add a bit of text to the commentary on line 1695 to point out that implementations may cache subsets of NAPOT regions, e.g., they're permitted to break NAPOT regions into 4KiB chunks in the TLBs.

gfavor commented on src/supervisor.tex:
```diff
@@ -1650,6 +1672,108 @@ \subsection{Addressing and Memory Protection}
+
+\begin{commentary}
+  This added step represents the creation of a single TLB entry covering the
+  entire NAPOT region.
```
See my preceding comment.

ingallsj commented on src/supervisor.tex:
```diff
@@ -1650,6 +1672,108 @@ \subsection{Addressing and Memory Protection}
+  (Valid) bit is clear.
+
+  If an implementation chooses to use a NAPOT PTE (or cached version thereof),
+  it may not consult the PTE specified by the algorithm in
```
"may not" is potentially misleading here: suggest "might not"

daniellustig replied:
Agreed, will fix, thanks.

daniellustig committed:
Address comments since last revision

daniellustig committed:
Merge branch 'master' into addr-trans-operational

daniellustig committed:
Various address translation updates

* Clarify translation algorithm atomicity requirements wrt SFENCE
* Allow A bit updates to be speculative (proposed recently, but not yet
  confirmed as a normative change we'd like to make)
* Justify translation algorithm atomicity wrt OS/supervisor stores
* Clarify "original" memory access in the translation algorithm
* Clarify "return to step 2" in the translation algorithm
* Pick a NAPOT encoding to start with.  This is only an attempt to make
  forward progress; it's not a final decision.

fintelia commented on src/supervisor.tex:
```diff
@@ -1407,9 +1414,21 @@ \subsection{Addressing and Memory Protection}
+In both cases, requiring atomicity ensures that the PTE update will not be
+interrupted by other S-mode writes to the page table, as such interruptions
+could lead to A/D bits being set on PTEs that are no longer valid.  Simple
+implementations may instead generate page-fault exceptions.
```
Does this need to say that the page-faults will only be triggered by non-speculative accesses?

gfavor commented on src/supervisor.tex:
```diff
@@ -1407,9 +1414,21 @@ \subsection{Addressing and Memory Protection}
+
+In both cases, requiring atomicity ensures that the PTE update will not be
+interrupted by other S-mode writes to the page table, as such interruptions
+could lead to A/D bits being set on PTEs that are no longer valid.  Simple
```
Isn't the range of "bad" possibilities broader, i.e. the A/D bits could be set on an arbitrarily new and different valid PTE (as well as an invalid PTE)?  (And these would also be very undesirables cases to allow to occur.)

gfavor commented on src/supervisor.tex:
```diff
@@ -1464,15 +1497,19 @@ \subsection{Virtual Address Translation Process}
+  \item Set $pte.a$ to 1 and, if the original memory access is a store, also set
+    $pte.d$ to 1.  Create a new address-translation cache entry mapping $b$ to
+    the value $pte$ for the address space identified in $\text{\em satp}.ASID$.
+  \item This update, the loading of $pte$ in step 2, and the creation of the
```
Why not explicitly specify the set of steps that must be part of this atomic operation (instead of just explicitly specifying the first and last steps and leaving it as an unstated implication that the intervening steps should also be included)?

gfavor commented on src/supervisor.tex:
```diff
@@ -1486,6 +1523,17 @@ \subsection{Virtual Address Translation Process}
+  The translation algorithm explicitly permits an arbitrary number of
+  translations to begin and then pause at any point for some arbitrary amount
+  of time before continuing.  This has two notable implications: 1) at any
+  time, the entire page table (i.e., all entries reachable by executing the
```
"Page table" refers to a specific 4KiB page table, whereas I assume you mean to refer to the whole tree of page tables pointed to by *atp?

gfavor commented on src/supervisor.tex:
```diff
@@ -1642,6 +1703,262 @@ \subsection{Addressing and Memory Protection}
+  designed to fit within the pre-existing Sv39 and Sv48 address translation
+  algorithms so as not to disrupt existing implementations or designs that
+  choose not to implement the scheme.
+
```
I think it would be worth adding "It is also allowable for a NAPOT PTE to be cached as any smaller size address mapping than the full NAPOT region size."  Not just as a explicit allowance to implementations, but also a heads up to software developers.

gfavor replied:
And I now see that this is touched on down below, but I still think it would be worth having this small sentence here as well.

daniellustig committed:
Another pass

* Relax the A/D update atomicity requirement
* Walk back some address-translation cache wording, for the sake of
  clarity
* In exchange, be a little more explicit about SFENCE ordering rules
* Do a bit of reorganizing
* Move the litmus tests to their own section, since they're not really
  NAPOT-specific
